### ISSUE TITLE
Multi-GPU training CNN hangs when using TensorboardCallback

### ISSUE THREAD

**Issue by ntdesilv** [fastai/fastai#3845]
Please confirm you have the latest versions of fastai, fastcore, and nbdev prior to reporting a bug (delete one): YES

**Describe the bug**
Hi! I have recently been experimenting with TensorBoard and fastai, especially as a means of tracking metrics in real-time with ClearML.

I've noticed that the steps to train a CNN across GPUs using Accelerate works fine, but the moment you introduce a `TensorBoardCallback` in training, it hangs indefinitely without any errors. Training the CNN on a multi-GPU instance but without distributed training (ie only using one of the GPUs and without including Accelerate) works perfectly fine too.

I mentioned this issue on the Accelerate repo: https://github.com/huggingface/accelerate/issues/900
And @muellerzr hypothesised that it's because TensorBoard can only run as a main process only which fastai doesn't guard. (Thanks, Zachary!)

**To Reproduce**
1. Spin up a notebook session with multiple GPUs. Here is the information on all of my settings here:
    ```Shell
    Accelerate version: 0.15.0
    OS: CentOS 7 (running JupyterLab through Docker with a CUDA-configured container
    Python version: 3.9.12
    numpy version: 1.23.5
    ClearML version: 1.8.2
    torch version:
    * torch==1.12.1+cu113
    * torchaudio==0.12.1+cu113
    * torchvision==0.13.1+cu113
    fastai version: 2.7.10
    protobuf version: 3.19.6 (because of tensorboard issues)
    accelerate configuration:
      * command_file: null
      * commands: null
      * compute_environment: LOCAL_MACHINE
      * deepspeed_config: {}
      * distributed_type: MULTI_GPU
      * downcast_bf16: 'no'
      * dynamo_backend: 'NO'
      * fsdp_config: {}
      * gpu_ids: all
      * machine_rank: 0
      * main_process_ip: null
      * main_process_port: null
      * main_training_function: main
      * megatron_lm_config: {}
      * mixed_precision: 'no'
      * num_machines: 1
      * num_processes: 4
      * rdzv_backend: static
      * same_network: true
      * tpu_name: null
      * tpu_zone: null
      * use_cpu: false
    CUDA version: 11.3
    EC2 instance type: p3.8xlarge
    ```
2. Take the following base script:
    ```python
    from fastai.vision.all import *
    
    from accelerate import notebook_launcher
    from fastai.distributed import *
    from clearml import Task, Logger
    from fastai.callback.tensorboard import TensorBoardCallback
    
    path = untar_data(URLs.PETS)/'images'
    
    # Not included - the credentials and host information for ClearML set as environment variables
    task = Task.init(project_name='Test Project', task_name='clearml-fastai-integration-demo-4')
    logger = Logger.current_logger()
    
    path = untar_data(URLs.PETS)/'images'
    task = Task.init(project_name='Listing Image Tagger', task_name='clearml-fastai-integration-demo-4')
    logger = Logger.current_logger()
    
    def train():
        print('Creating DataLoader')
        dls = ImageDataLoaders.from_name_func(
            path, get_image_files(path), valid_pct=0.2,
            label_func=lambda x: x[0].isupper(), item_tfms=Resize(224))
        print('Creating learner')
        learn = vision_learner(dls, resnet34, metrics=error_rate).to_fp16()
        print('Outside learn.distrib_ctx')
        with learn.distrib_ctx(in_notebook=True, sync_bn=False):
            print('Inside learn.distrib_ctx')
            # learn.fine_tune(2, cbs=[TensorBoardCallback()])
            learn.fine_tune(2)
    
    notebook_launcher(train, num_processes=4)
    ```
3. Paste it into a cell in a Jupyter Lab session
4. Uncomment one of the `learn.fine_tune` lines and comment the other (eg try first without any callbacks)
5. Run the cell
6. Swap the uncommented and commented `learn.fine_tune` lines (eg now try with `TensorBoardCallback`)

**Expected behavior**
I expect the model to train across GPUs with a `TensorBoardCallback` enabled.

**Error with full stack trace**
Here's the output when there is no callback:
```shell
Launching training on 4 GPUs.
Creating DataLoader
Creating DataLoader
Creating DataLoader
Creating DataLoader
Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

2022-12-02 04:02:37,332 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

2022-12-02 04:02:37,632 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
2022-12-02 04:02:38,745 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
Outside learn.distrib_ctx
2022-12-02 04:02:39,055 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
Outside learn.distrib_ctx
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
Outside learn.distrib_ctx
Outside learn.distrib_ctx
[W socket.cpp:401] [c10d] The server socket cannot be initialized on [::]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
Training Learner...
Inside learn.distrib_ctxInside learn.distrib_ctxInside learn.distrib_ctxInside learn.distrib_ctx



[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).

 0.00% [0/1 00:00<?]
epoch	train_loss	valid_loss	error_rate	time

 39.13% [9/23 00:06<00:09 0.0596]
```

Here's the output when using `TensorBoardCallback` as a callback:
```shell
Launching training on 4 GPUs.
Creating DataLoader
Creating DataLoader
Creating DataLoader
Creating DataLoader
Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

Creating learner
/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning:

The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.

/root/miniconda/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning:

Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet34_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet34_Weights.DEFAULT` to get the most up-to-date weights.

2022-12-02 03:58:35,509 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
2022-12-02 03:58:35,608 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
2022-12-02 03:58:35,646 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
2022-12-02 03:58:35,712 - clearml.model - INFO - Selected model id: b212faeef29d4a54861d19d9bd2a3bde
Outside learn.distrib_ctx
Outside learn.distrib_ctx
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
Outside learn.distrib_ctx
Outside learn.distrib_ctx
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:401] [c10d] The server socket cannot be initialized on [::]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
Training Learner...
Inside learn.distrib_ctx
Inside learn.distrib_ctxInside learn.distrib_ctx

Inside learn.distrib_ctx
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:558] [c10d] The client socket cannot be initialized to connect to [localhost]:29500 (errno: 97 - Address family not supported by protocol).
```

The output does not progress from here.

**Additional context**
To summarise:
| Using Accelerate to multi-process training? | Callbacks enabled? | Result |
|:---|:--:|---:|
| No | No | Runs successfully |
| No | Yes | Runs successfully |
| Yes | No | Runs successfully |
| Yes | Yes | Hangs indefinitely |
